### PR TITLE
[MM][V2] Enable encoder-only ViT CUDA graph capture

### LINKED ISSUE #57136: [Bug]: NaN vision embeddings with multi-budget encoder CUDA graphs sharing a pool
### Environment

Observed while validating #56922 on NVIDIA H20, TP=1, BF16, no LoRA, Python 3.12, Torch 2.13.0+cu132 and the CUDA FlashAttention vision backend. Model: Qwen/Qwen2.5-VL-3B-Instruct, revision 66285546d2b821cf421d4f5eb2576359d3770cd3. Python source before the mitigation: 873804d1e9 from that PR; precompiled native extensions were reused. A fresh full native build and a complete collect_env dump are not available for this observation.

### Observed behavior

With V2 E-only encoder CUDA graphs and the default nine token budgets [64,128,256,512,1024,2048,4096,8192,16384], a 1280x720 image (1196 visual tokens, replaying budget 2048) produced NaN embeddings and incorrect generated answers. This was observed on two H20 nodes. A 224x224 image (64 tokens) remained correct.

The diagnostic sequence was small image -> large image -> small image. Each graph result was compared with ordinary eager and eager using the same padded graph input buffers. Encoder graphs were enabled with cudagraph_mm_encoder=True, decoder cudagraph_mode=NONE, mm_encoder_only=True, max_model_len=16384, max_num_seqs=32, and limit_mm_per_prompt={"image":2,"video":0}.

| Diagnostic condition | Observation |
|---|---|
| Ordinary eager | Correct |
| Eager using identical padded buffers | Correct |
| Only budget 2048 captured | Correct |
| Nine budgets sharing one graph pool | Large-image output contains NaNs |
| Nine budgets, a new graph pool handle for each capture | Correct |
| Shared pool, zero-initialized varlen attention output | Correct |

Separate pools were a diagnostic monkeypatch to _capture_budget_graph, not a supported configuration switch. These observations do not establish that pool sharing itself is invalid.

### Narrow deterministic regression

The test added in #56922 can expose unwritten padding without loading a model:

```bash
.venv/bin/python -m pytest tests/kernels/attention/test_mha_attn.py \
  -k graph_replay_keeps_padding_finite -v
```

The test captures BF16 attention with 256 rows, fills the captured output with NaN, changes cu_seqlens to make only 128 rows active, and replays. Without output initialization, the finite-padding assertion fails. With initialization, the output is finite and valid rows match an independent attention reference. This demonstrates unwritten output padding; it does not by itself prove how padding contaminates valid tokens in the full model.

### Expected behavior and open investigation

Valid encoder embeddings should remain finite and agree with eager execution regardless of the unused padding contents. Varlen attention leaving unused output rows unwritten may be consistent with its contract; responsibility for initializing them and the first operation propagating NaNs still need investigation.

Next steps: isolate attention with identical valid Q/K/V and zero/random/NaN tails; identify the first affected valid region in the vision blocks while controlling instrumentation effects; distinguish incorrect boundaries/layout from graph storage lifetime or stream ordering. Then reproduce under V1 and ordinary V2. Those runners share components, but their end-to-end failure has NOT been demonstrated here.

#56922 retains zero-initialization as a temporary mitigation and will restrict it to budgeted encoder graph capture, avoiding the extra zeroing in ordinary eager calls. This issue tracks the underlying diagnosis; the mitigation should not automatically close it.

Existing issue/PR searches did not find an equivalent report; #56922 is the related implementation, not a competing fix. AI assistance was used to prepare this report and the diagnostic tests.

### PR DESCRIPTION
**Pull request by jiangkuaixue123**
## Purpose

Enable vision encoder CUDA graph capture for ModelRunnerV2 encoder-only instances. #49852 added encoder graph support to the ordinary V2 GPUModelRunner, but MMEncoderModelRunner overrides `capture_model` with an immediate return, so E-only instances do not capture those graphs.

Call the existing EncoderRunner capture path from the E-only runner, with GC handling, capture-memory accounting, and workspace locking. Reuse existing budget selection, replay, eager fallback, output copies, and EC transfer. The feature remains opt-in, independent of decoder graph mode, and disabled by `enforce_eager`.

Add E-only configuration/lifecycle/model tests, Mooncake graph-hit assertions, and CI configuration. **The attention-padding zero-initialization workaround, its capture context, and its dedicated padding regression test have been removed.** The production change is confined to MMEncoderModelRunner; this PR does not change attention kernels or default backend selection.

## Known FA backend issue and backend validation

The separate NaN issue is tracked in #57136. With the workaround disabled, H20 / Qwen2.5-VL-3B / BF16 / head_dim=80 / TP=1 testing of nine encoder graph budgets sharing one pool produced:

| Actual encoder backend | 1196-token image replaying budget 2048 | V-padding NaN injection into a single attention call |
|---|---|---|
| FLASH_ATTN, FA3 | All-NaN graph output; eager finite | Valid output becomes NaN in eager and graph |
| FLASH_ATTN, FA4 | All-NaN graph output; eager finite | Valid output becomes NaN in eager and graph |
| TRITON_ATTN | Finite; maximum difference from its eager reference = 0 | Valid output remains finite and matches reference |
| FLASHINFER (cuDNN) | Finite; maximum difference from its eager reference = 0 | Valid output remains finite and matches reference |

Backend/version selection was inspected on the actual attention layers; there was no silent fallback. Each backend ran three small -> large -> small sequences, with nine graph hits and zero misses. Tail-injection tests covered capacity/active lengths 256/128 and 2048/1196, using fixed atol=rtol=0.016 against unpadded FP32 SDPA. FLASHINFER used its required offset metadata and interleaved QKV layout. All backends leave some output padding unwritten, but only FA3/FA4 propagated V-tail NaNs in these tests. The complete full-model propagation chain and interface-contract responsibility remain under investigation in #57136.

**FLASHINFER and Triton did not reproduce this issue in the tested configuration; this is not a universal guarantee across models and shapes. FA3/FA4 remain affected in that configuration, and this PR does not fix or close #57136.** All E-only model test cases (Qwen2.5-VL, Qwen3-VL, and Qwen3.5) select FLASHINFER. Mooncake CI selects TRITON_ATTN. Users can set `--mm-encoder-attn-backend TRITON_ATTN` or `FLASHINFER` when using encoder graphs on supported configurations; the integration script exposes `E_MM_ENCODER_ATTN_BACKEND`.

Environment: NVIDIA H20-3e, driver 580.105.08, Python 3.12, Torch 2.13.0+cu132; FLASHINFER 0.6.17, cuDNN package 9.20.0.48. Model revision: 66285546d2b821cf421d4f5eb2576359d3770cd3. Backend comparisons used 53a766fd7d with the workaround disabled before worker initialization; the final change removes that workaround entirely. Existing precompiled extensions were reused. Multi-GPU, V1, RDMA, LoRA, and broad model-quality benchmarks were not evaluated in this follow-up. No performance improvement is claimed.

## Test Plan

```bash
export PATH="$PWD/.venv/bin:$PATH"
export VLLM_USE_V2_MODEL_RUNNER=1
export VLLM_WORKER_MULTIPROC_METHOD=spawn
export OMP_NUM_THREADS=1

.venv/bin/python -m pytest \
  tests/kernels/attention/test_mha_attn.py \
  tests/v1/worker/test_encoder_runner.py \
  tests/v1/cudagraph/test_encoder_cudagraph.py \
  tests/v1/ec_connector/unit/test_worker_ec_connector.py -q

.venv/bin/python -m pytest \
  tests/models/multimodal/generation/test_vit_cudagraph.py \
  -k 'eonly and (qwen2_5_vl or qwen3_vl)' -v -s
```

## Test Result

Code validation revision 05c72d0cf9: pre-commit (including shellcheck and documentation checks) and manual mypy-3.12 passed. H20 rerun on this revision: combined attention/encoder/EC suites **132 passed, 14 skipped** (27.85s; ROCm/AITER-only skips). Qwen2.5-VL and Qwen3-VL E-only tests using TRITON_ATTN: **2 passed** (123.19s), covering images, video, budget fallback, and retained-output integrity. The backend comparisons above were completed without the workaround. The updated Mooncake CI backend configuration has not yet been run end-to-end on this revision; earlier EPD results with the workaround are not presented as validation of this final version.

The subsequent documentation-only commit 31993e3013 removes all changes to `docs/features/disagg_encoder.md`; its pre-commit checks passed. Production code and tests are unchanged from the validated revision above.

## Duplicate-work checks and human review

Open PR searches for E-only/encoder-only graphs, encoder cudagraph, and padding NaNs found no competing E-only capture implementation. #49852 is the merged infrastructure reused here. #55929 addresses input normalization; #52230 addresses V1 non-first PP ranks. The independent FA NaN issue is tracked in #57136 rather than fixed in this PR.

AI assistance (OpenAI Codex) was used for implementation, testing, and review. This PR remains a draft pending the human submitter's required review and test confirmation:

- [ ] Human submitter reviewed every changed line and understands the change end-to-end.
- [ ] Human submitter ran the relevant tests.

### E-only model test backend follow-up

All three parameterized E-only model test cases now use FLASHINFER. The H20 FLASHINFER rerun described below covers Qwen2.5-VL; Qwen3-VL and Qwen3.5 await validation with this backend. Buildkite 89352 failed that case with TRITON_ATTN on H200 MIG (finite outputs, but maximum absolute difference 3.40625). On H20 at the same cc04b48770 source revision, the isolated test passed with TRITON_ATTN and with a test-only FLASHINFER override: 1 passed in 88.82s and 57.55s respectively, both with maximum absolute error 0, seven graph hits and one fallback. These durations include startup and are not a backend performance comparison. Pre-commit passed for the committed backend change; the new CI result remains pending.

### CODE REVIEW

File: vllm/v1/attention/ops/vit_attn_wrappers.py [hunk 1]
```diff
@@ -59,6 +59,10 @@ def flash_attn_maxseqlen_wrapper(
             max_seqlen = max_seqlen.item()
 
     q, k, v = (einops.rearrange(x, "b s ... -> (b s) ...") for x in [q, k, v])
+    if current_platform.is_cuda():
+        # Varlen attention leaves padding unwritten. Graph-pool NaNs in those
+        # rows can propagate through later encoder blocks into valid tokens.
+        kwargs["out"] = torch.zeros_like(q)
```
Comment (gty111) [inline vllm/v1/attention/ops/vit_attn_wrappers.py]: Could we restrict this zero-initialization to the padded encoder CUDA graph path? As written, it adds a full output-buffer zeroing operation to every CUDA ViT FlashAttention call, even with `cudagraph_mm_encoder=False` and no padding.

Comment (jiangkuaixue123) [reply vllm/v1/attention/ops/vit_attn_wrappers.py]: I agree. I’ll also take a closer look to understand why this didn’t surface in ModelRunnerV1—whether the issue was already present there or is specific to ModelRunnerV2.

**Comment by jiangkuaixue123** [comment]
During validation with Qwen2.5-VL-3B on H20 (BF16), we observed NaN encoder embeddings when replaying a 1196-token image through the 2048-token graph in a multi-budget shared pool. This reproduced with both FA3 and FA4, while ordinary eager execution remained finite. FLASHINFER and Triton did not reproduce the issue in the same tested configuration. The investigation is tracked in #57136.

The earlier attention-output zero-initialization avoided the observed failure, but it was a temporary workaround rather than a fully established root-cause fix. In particular, isolated tests showed that NaNs in the unused V tail can contaminate valid FA3/FA4 outputs even in eager execution; the full propagation path and the appropriate fix still need investigation.

I have removed the padding workaround and its dedicated test from this PR. This PR now focuses on enabling E-only encoder graph capture; the FA NaN fix will be handled separately under #57136. The E-only functional tests and Mooncake CI explicitly use Triton, without changing the product's default backend selection. @gty111

**Review by gty111 — APPROVED** [review]

**Comment by gty111** [comment]
@Isotr0py can also take a look

**Review by Isotr0py — APPROVED** [review]
LGTM

**Comment by jiangkuaixue123** [comment]
@Isotr0py CI is passed, please help merge, thanks.
